### ISSUE TITLE
Cassandra fails to start on EKS with "no permission to create directory /var/lib/cassandra/data"

### ISSUE THREAD

**Issue by jgoeres** [instaclustr/cassandra-operator#341]
Hi,

I have successfully tested v3.1.1 of the operator on my minikube environment. Now I tried to run it on EKS (1.14). Alas the pod goes into CrashLoopBackoff. The relevant log excerpts of the cassandra container:

    ERROR [main] ConcatenatedYamlConfigurationLoader.java:173 Unable to write cassandra configuration to /var/lib/cassandra-config.yaml
    java.nio.file.AccessDeniedException: /var/lib/cassandra/cassandra-config.yaml
            at sun.nio.fs.UnixException.translateToIOException(UnixException.java:84) ~[na:1.8.0_232]
            at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102) ~[na:1.8.0_232]
            at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107) ~[na:1.8.0_232]
            at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214) ~[na:1.8.0_232]
            at java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:434) ~[na:1.8.0_232]
            at java.nio.file.Files.newOutputStream(Files.java:216) ~[na:1.8.0_232]
            at java.nio.file.Files.write(Files.java:3292) ~[na:1.8.0_232]
            at com.instaclustr.cassandra.k8s.ConcatenatedYamlConfigurationLoader.loadConfig(ConcatenatedYamlConfigurationLoader.java:171) ~[cassandra-k8s-addons-1.2.0.jar:1.2.0]
            at org.apache.cassandra.config.DatabaseDescriptor.loadConfig(DatabaseDescriptor.java:276) [apache-cassandra-3.11.5.jar:3.11.5]
            at org.apache.cassandra.config.DatabaseDescriptor.daemonInitialization(DatabaseDescriptor.java:152) [apache-cassandra-3.11.5.jar:3.11.5]
            at org.apache.cassandra.config.DatabaseDescriptor.daemonInitialization(DatabaseDescriptor.java:137) [apache-cassandra-3.11.5.jar:3.11.5]

and the log ends with

    INFO  [main] StartupChecks.java:176 JMX is enabled to receive remote connections on port: 7199
    INFO  [main] SigarLibrary.java:44 Initializing SIGAR library
    INFO  [main] SigarLibrary.java:180 Checked OS settings and found them configured for optimal performance.
    WARN  [main] StartupChecks.java:332 Directory /var/lib/cassandra/data doesn't exist
    ERROR [main] CassandraDaemon.java:759 Has no permission to create directory /var/lib/cassandra/data

So it seems there is a permission problem with the data volume mounted into /var/lib/cassandra.

I saw that the ownership of that folder is set to 999 in the "install-cassandra" script used in the Dockerfile. I wonder which permissions take precedence if at runtime an external volume is mounted into this location. Could it be that the permission change should have to be done after the volumes are mounted, i.e., at some place in the start script, before stepping down from root?
I remember seeing something like this when I an inspected an older version of an elasticsearch image (5.x something). They basically changed the ownership of their data directory etc. to the user they had created in the docker file and only then stepped down from root to start ES itself.

Today, their entrypoint looks a bit different, but it basically seems to be doing the same thing: 
https://github.com/elastic/dockerfiles/blob/v7.5.2/elasticsearch/bin/docker-entrypoint.sh

What I can say for sure is that this very ES image is working fine in the same EKS cluster (and on minikube) with the same storage class. I am using the "Elastic Cloud on Kubernetes" operator to create the ES cluster.

This is the CassandraDataCenter spec I am using:

    apiVersion: cassandraoperator.instaclustr.com/v1alpha1
    kind: CassandraDataCenter
    metadata:
      name: my-cassandra
      labels:
        app: cassandra
    spec:
      nodes: 1
      cassandraImage:  <myEcrRegistryID>.dkr.ecr.eu-central-1.amazonaws.com/cassandra-operator/cassandra-3.11.5:v3.1.1
      sidecarImage: <myEcrRegistryID>.dkr.ecr.eu-central-1.amazonaws.com/cassandra-operator/cassandra-sidecar:v3.1.1
      imagePullPolicy: "IfNotPresent"
      imagePullSecrets:
        - name: registrysecret
      resources:
        limits:
          memory: 1024Mi
        requests:
          memory: 1024Mi
      sidecarResources:
        limits:
          memory: 512Mi
        requests:
          memory: 512Mi
    
      dataVolumeClaimSpec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 10Gi
      optimizeKernelParams: true
      serviceAccountName: cassandra
    
Note that I am not specifying a storageClass in the dataVolumeClaimSpec here, but just use the EKS default storage class (gp2,i.e., an EBS disk). On minikube, I am also using the default storage class, there it is a hostPath volume.

Regards

J

**Comment by smiklosovic** [comment]
Hi @jgoeres ,

have you tried this? I know it is for Azure but I suspect this is the case of AWS as well, notice there is "fsGroup: 999" in `spec`. It means that, shortly, mounted volume will behave as if it is all on 999 so it will be writable.

Docker containers are meant to be run with `cassandra` user all the time, we are not stepping down / up to `root` anywhere and it was meant to be like like that.

https://github.com/instaclustr/cassandra-operator/blob/master/doc/providers/aks.md#requirements


Maybe unrelated but if you plan to run your cluster in a multi-rack deployment, it might happen that your PV is in a different region from a node itself so it will not be able to mount it. This is the workaround for that: https://github.com/instaclustr/cassandra-operator/blob/master/doc/persistent_volumes.md

**Comment by jgoeres** [comment]
Hi,

thank you, with the fsGroup setting it is starting up fine now. 
I will switch to the performance service account ( I noticed the different PodSecurityPolicies and service accounts and was wondering what difference it makes, but hadn't tried it), thanks.

Also thanks für pointing out the multi-zone problem. I actually ran into that issue when we worked on deploying another service (one developed in-house) across multiple zones on AWS. That was even before before Kubernetes 1.13 or 1.14(?) when the "WaitForFirstConsumer" "volumeBindingMode" didn't even exist yet. My workaround was to create dedicated storage classes in the helm chart for each AWS zone and using them in the statefulset for each zone.  Problem is that the default setting of my Helm chart will use the default storage class of the cluster (since it requires the minimal amount of configuration by the Helm user). Unfortunately, on EKS, the default "gp2" storage class doesn't have "volumeBindingMode" set to "WaitForFirstConsumer". While we currently only plan to use this Helm chart for our SaaS hosting where we handle the operations, it will eventually also be used by customers directly, where this will be another pitfall.
